File: ripbull-chat-sdk/src/main/java/com/ripbull/coresdk/core/base/BaseRepository.java
package com.ripbull.coresdk.core.base;

import android.text.TextUtils;
import androidx.annotation.NonNull;
import com.annimon.stream.function.Function;
import com.annimon.stream.function.Supplier;
import com.ripbull.ertc.cache.database.DataSource;
import com.ripbull.ertc.cache.preference.PreferenceManager;
import com.ripbull.ertc.remote.NetworkManager;
import com.ripbull.coresdk.data.DataManager;

/** Created by DK on 2019-05-04. */
public class BaseRepository {

  private final DataManager dataManager;

  protected BaseRepository(DataManager dataManager) {
    this.dataManager = dataManager;
  }

  protected DataManager data() {
    return dataManager;
  }

  protected DataSource db() {
    return dataManager.db();
  }

  protected NetworkManager network() {
    return dataManager.network();
  }

  protected PreferenceManager preference() {
    return dataManager.preference();
  }

  protected String getTenantId() {
    return data().preference().getTenantId();
  }

  protected String getUserId() {
    return data().preference().getUserId();
  }

  protected String getAppUserId() {
    return data().preference().getAppUserId();
  }

  protected String getApiKey() {
    return data().preference().getApiKey();
  }

  protected String getChatUserId() {
    return data().preference().getChatUserId();
  }

  protected String getName() {
    return data().preference().getName();
  }

  protected String getFCMToken() {
    return data().preference().getFcmToken();
  }

  protected String getDeviceId() {
    return data().preference().getDeviceId();
  }

  @NonNull
  protected <R> R withTenant(Function<String, R> function, Supplier<R> errorSupplier) {
    final String tenantId = data().preference().getTenantId();
    if (tenantId != null && !TextUtils.isEmpty(tenantId)) {
      return function.apply(tenantId);
    }
    return errorSupplier.get();
  }
}

File: InAppChat/src/main/java/io/inappchat/inappchat/core/base/BaseRepository.kt
package io.inappchat.inappchat.core.base

import android.text.TextUtils
import io.inappchat.inappchat.cache.database.DataSource
import io.inappchat.inappchat.cache.preference.PreferenceManager
import io.inappchat.inappchat.data.DataManager
import io.inappchat.inappchat.remote.NetworkManager

/** Created by DK on 2019-05-04.  */
open class BaseRepository protected constructor(dataManager: DataManager) {
    private val dataManager: DataManager

    init {
        this.dataManager = dataManager
    }

    protected fun data(): DataManager {
        return dataManager
    }

    protected fun db(): DataSource {
        return dataManager.db()
    }

    protected fun network(): NetworkManager {
        return dataManager.network()
    }

    protected fun preference(): PreferenceManager {
        return dataManager.preference()
    }

    protected val tenantId: String
        protected get() = data().preference().tenantId!!
    protected val userId: String
        protected get() = data().preference().userId!!
    protected val appUserId: String
        protected get() = data().preference().appUserId!!
    protected val apiKey: String
        protected get() = data().preference().apiKey!!
    protected open val chatUserId: String
        protected get() = data().preference().chatUserId!!
    protected val name: String
        protected get() = data().preference().name!!
    protected val fcmToken: String
        protected get() = data().preference().fcmToken!!
    protected open val deviceId: String?
        protected get() = data().preference().deviceId

    protected fun <R> withTenant(function: (String) -> R, errorSupplier: () -> R): R {
        val tenantId = data().preference().tenantId
        return if (tenantId != null && !TextUtils.isEmpty(tenantId)) {
            function(tenantId)
        } else errorSupplier()
    }
}